src/backend/parser/parse_expr.c

```diff
@@ -2325,6 +2328,18 @@ transformCoalesceExpr(ParseState *pstate, CoalesceExpr *c)
 	{
 		Node	   *e = (Node *) lfirst(args);
 		Node	   *newe;
+		Oid		   etype = exprType(e);
+
+		/*
+		 *	T-SQL treats constant string literals as VARCHAR. Hence,
+		 *	coercing into VARCHAR before coercing it to the common type.
+		 */
+		if (sql_dialect == SQL_DIALECT_TSQL && !c->tsql_is_null && 
+			etype == UNKNOWNOID && IsA(e, Const) &&
+			handle_constant_literals_hook)
```

Comment:
Let's only check for dialect and `handle_constant_literals_hook` here. The rest of the things should be moved inside the hook implementation. This will reduce overhead on a regular PG query